datafusion/core/src/execution/context.rs

```diff
@@ -1557,14 +1557,6 @@ impl SessionState {
                 .register_catalog(config.default_catalog.clone(), default_catalog);
         }
 
-        let optimizer_config = OptimizerConfig::new().filter_null_keys(
```

Comment:
Moving this into the `OptimizerConfig` passed at optimize time is the fix for #4638